(#666) Added unit tests for CallableOf

llorllale opened the pull request:
As per #666:

Added missing tests for these ctors:
* `CallableOf(final Runnable runnable)`
* `CallableOf(final Proc<X> proc)`
* `CallableOf(final Func<X, T> fnc, final X ipt)`

0crat commented:
Job [#691](https://github.com/yegor256/cactoos/issues/691) is now in scope, role is `REV`

<!-- https://www.0crat.com/footprint/C63314D6Z/511420606, version: 1.0-SNAPSHOT, hash: 038723b0 -->

0crat commented:
This pull request [#691](https://github.com/yegor256/cactoos/issues/691) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C63314D6Z/511422664). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

<!-- https://www.0crat.com/footprint/C63314D6Z/511423671, version: 1.0-SNAPSHOT, hash: 038723b0 -->

amihaiemil commented on src/test/java/org/cactoos/func/CallableOfTest.java:
```diff
@@ -47,4 +50,39 @@ public void convertsFuncIntoCallable() throws Exception {
+    public void convertsRunnableIntoCallable() throws Exception {
+        final AtomicBoolean flag = new AtomicBoolean(false);
+        new CallableOf<>(
+            (Runnable) () -> { flag.set(true); }
```
@llorllale I personally don't like these casts (even though you said they are for readability), because it's clear enough from the test's name what it does. Plus, the combination with the lambda syntax looks weird, some may thing that the cast is actually necessary. WDYT?

llorllale replied:
@amihaiemil ok, I'll remove the cast

amihaiemil commented on src/test/java/org/cactoos/func/CallableOfTest.java:
```diff
@@ -47,4 +50,39 @@ public void convertsFuncIntoCallable() throws Exception {
+    public void convertsProcIntoCallable() throws Exception {
+        final AtomicBoolean flag = new AtomicBoolean(false);
+        new CallableOf<>(
+            (Proc<AtomicBoolean>) (unused) -> { flag.set(true); }
```
@llorllale What is this ``unused``? What is its type and how could we make sure that ``CallableOf`` doesn't change it before passing it to the Proc's method?

llorllale replied:
@amihaiemil I'm actually not sure what the purpose of having a ctor accepting a `Proc` is.

`Proc` is a functional interface with its SAM like this:
```java
void exec(X input) throws Exception;
```

`CallableOf` is a `Callable` and thus its `call()` takes no arguments. I can't think of a decent way to make this test make any more sense.

amihaiemil replied:
@llorllale hmm, yes, I see ``CallableOf`` wraps it inside FuncOf:  ``this(new FuncOf<>(proc));`` -- then we should see what FuncOf does with the Proc, if it uses that parameter somehow. If not, then I think it should be taken out.

We can leave if like this and maybe you can open an issue and ask clarifications? (since you noticed this in the first place)

amihaiemil commented on src/test/java/org/cactoos/func/CallableOfTest.java:
```diff
@@ -47,4 +50,39 @@ public void convertsFuncIntoCallable() throws Exception {
+    public void convertsFuncWithInputIntoCallable() throws Exception {
+        MatcherAssert.assertThat(
+            new CallableOf<>(
+                (Func<Integer, Integer>) num -> num + 1,
```
@llorllale Can you add accolades here? ``num -> { num + 1; }`` so we are consistent in all tests.

llorllale replied:
@amihaiemil adding the brackets there will then force the expression to be changed to `num -> { return num + 1; }`. Is that what you want?

amihaiemil replied:
@llorllale Right, then leave it like that. My bad, sorry, still hating the Java 8 syntax :(

amihaiemil commented:
@llorllale And do you know what happens if the encapsulated Runnable throws an exception? Will it bubble up or be silently swallowed? Maybe write a test for that too -- if it fails, ignore it and add a puzzle, someone should fix the situation later.

amihaiemil commented:
@llorllale Oh, and change the PRs title, to ``Added unit tests for CallableOf`` -- currently the title sounds like the issue, not the solution for it.

llorllale commented:
@amihaiemil as per discussion:

* Fixed lambda formatting/removed unnecessary casts
* Added new test case for checking if runtime exceptions from `Runnable` are wrapped

amihaiemil commented on src/test/java/org/cactoos/func/CallableOfTest.java:
```diff
@@ -54,19 +52,27 @@ public void convertsFuncIntoCallable() throws Exception {
     }
 
+    @SuppressWarnings("PMD.AvoidThrowingRawExceptionTypes")
+    @Test(expected = Exception.class)
```
@llorllale Well, ``Exception`` here doesn't guarantee that it's the one from inside the Runnable, right? Maybe ``throw IllegalStateException ("from runnable")`` and here ``expected = IllegalStateException.class)`` -- this way you can also get rid of the PMD warning.

llorllale replied:
@amihaiemil is the intention here that `CallableOf` wraps the runtime exception into a checked exception or not?

amihaiemil replied:
@llorllale Ah, ``public T call() throws Exception`` -- very poor choice of design. When I see ``throws Exception`` I read it as "programmer had no idea how to handle exceptions"... I think this one is even from ``Callable``, so JDK. Nice; then leave the ``expected`` like that. But you can still throw ``IllegalStateException`` to get rid of the warning, I think, no?

llorllale replied:
@amihaiemil ok, switching to `IllegalStateException`

llorllale commented:
@amihaiemil as per discussion
* Switched to using `IllegalStateException` in order to avoid having to suppress that PMD error
* Fixed PR title

please review

amihaiemil commented:
@rultor good to merge

rultor commented:
> @rultor good to merge

@amihaiemil Thanks for your request. @yegor256 Please confirm this.

llorllale commented:
ping @yegor256

amihaiemil commented:
@yegor256 ping

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/13958-367714287)

rultor commented:
> @rultor merge

@yegor256 Done! FYI, the full log is [here](http://www.rultor.com/t/13958-367714287) (took me 11min)

0crat commented:
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job, as in [§30](http://datum.zerocracy.com/pages/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/C63314D6Z/531545450, version: 1.0-SNAPSHOT, hash: bcb90c38 -->

ypshenychka commented:
@0crat quality good

0crat commented:
`QA` review completed: +15 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka), total is [+630](https://www.0crat.com/u/ypshenychka)

<!-- https://www.0crat.com/footprint/C63314D6Z/540717895, version: 1.0-SNAPSHOT, hash: b913e8dc -->

0crat commented:
Order was finished, quality was "good": +20 points just awarded to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), total is [+945](https://www.0crat.com/u/amihaiemil)

<!-- https://www.0crat.com/footprint/C63314D6Z/540720967, version: 1.0-SNAPSHOT, hash: b913e8dc -->

0crat commented:
> @0crat quality good ([here](https://github.com/yegor256/cactoos/issues/691#issuecomment-367927407))

@ypshenychka You can't do that, unless you have one of these roles: `ARC`, `PO`. Your current roles are: `QA`.

<!-- https://www.0crat.com/footprint/C63314D6Z/540721988, version: 1.0-SNAPSHOT, hash: b913e8dc -->

0crat commented:
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +15 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+10794](https://www.0crat.com/u/yegor256)

<!-- https://www.0crat.com/footprint/C63314D6Z/540723004, version: 1.0-SNAPSHOT, hash: b913e8dc -->

0crat commented:
This pull request [#691](https://github.com/yegor256/cactoos/issues/691) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol), here is [why](https://www.0crat.com/footprint/C63314D6Z/591812558). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

<!-- https://www.0crat.com/footprint/C63314D6Z/591812573, version: 1.0-SNAPSHOT, hash: 2a60d117 -->

neonailol commented:
@yegor256 i don't understand why this assigned to me, pr is closed and merged

yegor256 commented:
@llorllale refuse

yegor256 commented:
@0crat refuse

yegor256 commented:
@0crat out

0crat commented:
> @0crat refuse ([here](https://github.com/yegor256/cactoos/issues/691#issuecomment-369564662))

@yegor256 The user @neonailol[/z](https://www.0crat.com/u/neonailol) resigned from [#691](https://github.com/yegor256/cactoos/issues/691), please stop working. Reason for job resignation: Order was cancelled

<!-- https://www.0crat.com/footprint/C63314D6Z/601137225, version: 1.0-SNAPSHOT, hash: 3b2de08b -->

0crat commented:
> @0crat out ([here](https://github.com/yegor256/cactoos/issues/691#issuecomment-369564676))

@yegor256 The job [#691](https://github.com/yegor256/cactoos/issues/691) is now out of scope

<!-- https://www.0crat.com/footprint/C63314D6Z/601137238, version: 1.0-SNAPSHOT, hash: 3b2de08b -->

0crat commented:
> @0crat out ([here](https://github.com/yegor256/cactoos/issues/691#issuecomment-369564676))

@yegor256 Job [#691](https://github.com/yegor256/cactoos/issues/691) was not in scope

<!-- https://www.0crat.com/footprint/C63314D6Z/601138254, version: 1.0-SNAPSHOT, hash: 3b2de08b -->
